Expose dbus_register and dbus_unregister vfuncs

swsnr opened the pull request:
Using this vfunc makes exporting objects on the app's bus name much easier and more reliable.

I've rewritten the corresponding example to use these two methods, as the GIO docs recommend to use these vfuncs for object registration.  Also, registering an object in startup is prone to race conditions wrt to bus ownership as the object gets registered only after the application claimed the bus name.

swsnr commented:
This is my first time doing glib FFI, so I'd ask for a careful review, especially around ownership of pointers.

If this is acceptable, I'd very much appreciate a timely backport and release, as I need this to reliably export an object on the bus in my code.

swsnr commented:
Mmh, did I cause this CI error?

sdroege commented:
> Mmh, did I cause this CI error?

No, the docs job is just randomly failing.

sdroege commented on gio/src/subclass/application.rs:
```diff
@@ -110,6 +110,10 @@ pub trait ApplicationImpl:
         self.parent_handle_local_options(options)
     }
+
+    fn dbus_register(&self, connection: &DBusConnection, object_path: &str) -> Result<bool, Error> {
```
This should return `Result<(), Error>` AFAICS. The C function is supposed to either return `FALSE` and set the error, or return `TRUE` and not set the error.

swsnr replied:
Thanks, I'll change this. I must admit that I didn't entirely understand the semantics here 😇

sdroege replied:
Me neither, the documentation was confusing and I had to look at the implementation :)

swsnr replied:
😅

sdroege commented on gio/src/subclass/application.rs:
```diff
@@ -390,6 +425,30 @@ unsafe extern "C" fn application_handle_local_options<T: ApplicationImpl>(
+        &glib::GString::from_glib_borrow(object_path),
+    ) {
+        Ok(true) => glib::ffi::GTRUE,
+        Ok(false) => glib::ffi::GFALSE,
```
And same here

sdroege commented on examples/gio_dbus_register_object/main.rs:
```diff
@@ -94,24 +97,33 @@ mod imp {
-                                println!("{greet}");
-                                Ok(Some(greet.to_variant()))
+                .invoke_and_return_future_local(glib::clone!(
+                    #[strong(rename_to = app)]
```
I think this causes a reference cycle. Should probably be a weak reference at least

swsnr replied:
Sure, can change this, though I'm not really following where the cycle would come from, or why it'd be harmful given that the app usually lives for the entire lifetime of the process 🤔

But I'm also only superficially familiar with all this, so I trust your judgement 😅

sdroege reviewed (commented):
Looks good to me otherwise, thanks :)

bilelmoussaoui commented on gio/src/subclass/application.rs:
```diff
@@ -286,6 +339,8 @@ unsafe impl<T: ApplicationImpl> IsSubclassable<T> for Application {
         klass.shutdown = Some(application_shutdown::<T>);
         klass.startup = Some(application_startup::<T>);
         klass.handle_local_options = Some(application_handle_local_options::<T>);
+        klass.dbus_register = Some(application_dbus_register::<T>);
```
please add name_lost as well while you are at it, that would leave only add_platform_data missing which is not really useful as we don't have GVariantBuilder support

swsnr replied:
Okay, will do.

swsnr commented:
@bilelmoussaoui @sdroege I think I addressed everything.  I added name_lost too; I hope I did it right, because I didn't really understand what the return value does and it doesn't seem to be documented :shrug:

bilelmoussaoui commented on gio/src/subclass/application.rs:
```diff
@@ -390,6 +462,54 @@ unsafe extern "C" fn application_handle_local_options<T: ApplicationImpl>(
+) -> glib::ffi::gboolean {
+    let instance = &*(ptr as *mut T::Instance);
+    let imp = instance.imp();
+    if imp.name_lost() {
```
Use imp.name_lost().into_glib() directly

swsnr replied:
Yeah, right :see_no_evil: I'll also fix the other conditional that broke the build, but probably not before tomorrow evening; it's getting late here :slightly_smiling_face:

swsnr replied:
Fixed.

swsnr commented on gio/src/subclass/application.rs:
```diff
@@ -266,6 +278,63 @@ pub trait ApplicationImplExt: ApplicationImpl {
+            let parent_class = data.as_ref().parent_class() as *mut ffi::GApplicationClass;
+            let f = (*parent_class)
+                .name_lost
+                .expect("No parent class implementation for \"name_lost\"");
```
Is this correct?  Most code around here uses `.expect` to access the parent function, but `handle_local_option` properly unpacks the `Option`.  I'm not sure when to use either, so I'd appreciate if you'd check this.

bilelmoussaoui replied:
Unfortunately you have to check the C code and see if the Object has a default implementation, meaning there would always be a function to chain up.

sdroege replied:
Not only that, but also whether this function is mandatory. It might not have a default implementation but the caller might just call it always without checking if there's one, in which case it's mandatory

sdroege replied:
Checking the code, it looks mandatory so unwrapping/expect here is OK.

bilelmoussaoui replied:
I am curious though if this should return one of the two glib enums we have that internally maps to a boolean type

sdroege replied:
Return value should be a `glib::Propagation` I think

swsnr replied:
@sdroege Thanks for checking and saving me the trip to C 🙏

I'll change the return value to `Propagation`

sdroege commented on gio/src/subclass/application.rs:
```diff
@@ -266,6 +270,33 @@ pub trait ApplicationImplExt: ApplicationImpl {
+                &mut err,
+            );
+            if res == glib::ffi::GFALSE {
+                debug_assert!(!err.is_null());
```
That assertion is already in `from_glib_full()`, but you might want to check the in the `TRUE`-case it is always `NULL`

swsnr commented on examples/gio_dbus_register_object/main.rs:
```diff
@@ -94,24 +97,33 @@ mod imp {
+                                    Ok(Some(greet.to_variant()))
+                                }
+                                HelloMethod::GoodBye => {
+                                    app.inspect(|app| app.quit());
```
Mmh, isn't the whole point and purpose of `inspect` in contrast to `map` to run side effects?  I always use it that war 😅

But I'm fine to follow whatever style you prefer so I'll change it this evening.

swsnr reviewed (commented):
I've addressed all remarks, changed the return value of `name_lost` to `Propagation`, and rebased onto main.

swsnr commented:
While I was at it I've pushed another commit which replaces the IOErrorEnum's in the DBus example with `DBusError` which seemed more appropriate.
